src/GameLogic/PlugIns/InvasionEvents/HappyHourPlugIn.cs

```diff
@@ -0,0 +1,220 @@
+
+                    await this.OnFinishedAsync(state).ConfigureAwait(false);
+
+                    logger.LogInformation($"event finished");
+
+                    break;
+                }
+
+            default:
+                throw new NotImplementedException("Unknown state.");
+        }
+    }
+
+    /// <inheritdoc />
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "VSTHRD100:Avoid async void methods", Justification = "Catching all Exceptions.")]
+    public virtual async void ObjectAddedToMap(GameMap map, ILocateable addedObject)
+    {
+    }
+
+    /// <summary>
+    /// Get a unique state per GameContext.
+    /// </summary>
+    /// <param name="gameContext">GameContext.</param>
+    protected GameServerState GetStateByGameContext(IGameContext gameContext)
+    {
+        var type = this.GetType();
+
+        var statesPerType = _states.GetOrAdd(type, newType => new() { });
+
+        return statesPerType.GetOrAdd(gameContext, gameCtx => new GameServerState(gameContext));
+    }
+
+    /// <summary>
+    /// Returns true if the player stays on the map.
+    /// </summary>
+    /// <param name="player">The player.</param>
+    protected bool IsPlayerOnMap(Player player)
+    {
+        var state = this.GetStateByGameContext(player.GameContext);
+
+        return player.CurrentMap is { } map
+            && player.PlayerState.CurrentState != PlayerState.Disconnected
+            && player.PlayerState.CurrentState != PlayerState.Finished;
+    }
+
+    /// <summary>
+    /// Send a golden message to player's client.
+    /// </summary>
+    /// <param name="player">The player.</param>
+    protected async Task TrySendStartMessageAsync(Player player)
+    {
+        var configuration = this.Configuration;
+
+        if (configuration is null)
+        {
+            return;
+        }
+
+        var message = configuration.Message ?? "Happy Hour event started!";
+
+        if (this.IsPlayerOnMap(player))
+        {
+            try
+            {
+                await player.InvokeViewPlugInAsync<IShowMessagePlugIn>(p => p.ShowMessageAsync(message, Interfaces.MessageType.GoldenCenter)).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                player.Logger.LogDebug(ex, "Unexpected error sending start message.");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Calls after the state changed to Prepared.
+    /// </summary>
+    /// <param name="state">State.</param>
+    protected virtual async ValueTask OnPreparedAsync(GameServerState state)
+    {
+        await state.Context.ForEachPlayerAsync(p => this.TrySendStartMessageAsync(p)).ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Calls after the state changed to Started.
+    /// </summary>
+    /// <param name="state">State.</param>
+    protected virtual async ValueTask OnStartedAsync(GameServerState state)
+    {
+        state.Context.Configuration.ExperienceRate = state.NewExperienceRate;
```

Comment:
@sven-n

I know several ways:
- GameConfiguration
- MapDefiniton
- Character expRate attribute

Which one more suitable for this task?